### PR TITLE
studio-followers addon

### LINKED ISSUE #2940: Invite followers to studios
### Is your request related to a problem? Please describe.

Since scratch's studio update, they've quietly removed the ability to invite your followers. I think that's stupid.

### Describe the solution you'd like

I suggest someone make an addon that brings back that ability.

### PR DESCRIPTION
**Pull request by FunctionalMetatable**
Resolves #2940

### Changes

Adds a studio followers modal

### Reason for changes

That was a great feature

### Tests

Tested locally

image

### CODE REVIEW

File: addons/studio-followers/addon.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "Add followers to your studios",
+  "description": "Creates a modal that you can add your followers to a studio.",
```
Comment (pufferfish101007) [inline addons/studio-followers/addon.json:3]: I think 'Creates a modal that you can use to invite your followers to studios.' might work better both grammatically and logically.

File: addons/studio-followers/addon.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "credits": [
+    {
+      "name": "9gr",
+      "link": "https://scratch.mit.edu/users/9gr"
+    }
+  ],
+  "userscripts": [
+    {
+      "url": "userscript.js",
+      "matches": ["https://scratch.mit.edu/studios_www/*"],
```
Comment (pufferfish101007) [inline addons/studio-followers/addon.json:13]: This and the other url are incorrect - please replace both with "studios".

Comment (FunctionalMetatable) [reply addons/studio-followers/addon.json:13]: Isn't this needed for the scratchr2 migration thing?

Comment (apple502j) [reply addons/studio-followers/addon.json]: Nope. In fact this wouldn't work.

Comment (FunctionalMetatable) [reply addons/studio-followers/addon.json]: Oops, I forgot to update my fork before making a PR

**Comment by WorldLanguages** [comment]
Should also let you add by "following"
And we need to make sure this is compatible with @mxmou's "old studio layout" theme
If so, maybe we can merge today for v1.17.0

**Comment by GrahamSH-LLK** [comment]
Why does the modal header have so much radius on all sides? That looks weird

**Comment by FunctionalMetatable** [comment]
> Should also let you add by "following"

People that you're following are people that normally don't want to be invited to studios by you (unless they follow you too). I think that this kind of addon is only suitable for follower inviting, not adding following as well

**Comment by WorldLanguages** [comment]
> People that you're following are people that normally don't want to be invited to studios by you (unless they follow you too). I think that this kind of addon is only suitable for follower inviting, not adding following as well

Sure, I'm simply suggesting this because old studios had this feature too.

**Comment by WorldLanguages** [comment]
@FunctionalMetatable Please give me a few minutes to make a few changes to this PR.

**Comment by WorldLanguages** [comment]
I'm still not done

**Comment by FunctionalMetatable** [comment]
> > People that you're following are people that normally don't want to be invited to studios by you (unless they follow you too). I think that this kind of addon is only suitable for follower inviting, not adding following as well
> 
> Sure, I'm simply suggesting this because old studios had this feature too.

i completely forgot about the old studios that had the following, it seems a long time ago but it left like last week or smth

**Comment by WorldLanguages** [comment]
@FunctionalMetatable Tell me if I missed something, but there was a lot of unneeded and slightly incorrect CSS

**Comment by WorldLanguages** [comment]
@mxmou Please approve commit above ^ I tried my best :P

**Comment by WorldLanguages** [comment]
TODOs:

- [ ] Only request followers when modal is active
- [ ] Add following (only request when active as well)
- [ ] Do infinite scrolling instead of requesting all followers
- [ ] mxmou approval

Hopefully merge today, this is perfect for v1.17.0 release

**Comment by WorldLanguages** [comment]
@FunctionalMetatable Remember locales :P
And probably make it `Hmm, it seems like you are not allowed to do that. Make sure you're logged in.` so that translators can use the same string as studio-tools

**Comment by mxmou** [comment]
> @mxmou Please approve commit above ^ I tried my best :P

I like the way it looks with old-studio-layout, but there are some issues unrelated to that:
- Why is the button not next to the invite row?
- For consistency, it should also be possible to close the modal by clicking outside or using a "Done" button.
- In most other parts of Scratch, profile pictures are square and stretch to fit.

**Comment by WorldLanguages** [comment]
@mxmou Kinda unrelated. For the "old studio layout" addon, can we visually keep the "browse projects" and "invite followers" buttons where they would be, and have clicking them open the bottom bar (the real button?)
I fear some users might enable old studio layout and not find how to add projects or invite followers (the bottom bar can be easily ignored by the brain). We would use a userscript to add a button in the same place as the real one would be if we didn't apply position:fixed to it (and use `addon.tab.displayNoneWhileDisabled` to hide it with dynamicDisable)

**Comment by WorldLanguages** [comment]
@mxmou 
I suggest:
- Moving the "invite followers" button next to the invite by username button:
image
With a similar separator to:
image
- Add a "done" button to this modal
- Close this modal when clicking outside
- Square pfps and stretch to fit
- Do not forget infinite-scroll's "fix footer" feature, which also uses the bottom bar. What should we do for those cases?

**Comment by WorldLanguages** [comment]
The active tab (followers/following) should have class "active"

**Comment by WorldLanguages** [comment]
@FunctionalMetatable Try adding the "done" button. To get the word "done", use `addon.tab.scratchMessage("general.done")`

**Comment by FunctionalMetatable** [comment]
> @FunctionalMetatable Try adding the "done" button. To get the word "done", use `addon.tab.scratchMessage("general.done")`

can't do this rn as im gonna go to bed now (if you'd like I can do this tomorrow?)

**Comment by FunctionalMetatable** [comment]
can't do this now as I'm going to bed now, (if you'd like I can do it tomorrow?)

**Comment by WorldLanguages** [comment]
@FunctionalMetatable It's fine, I did not mean to make that comment a demand :P

**Comment by WorldLanguages** [comment]
I'm going to make edits to this PR in a few minutes

**Comment by GrahamSH-LLK** [comment]
> I'm going to make edits to this PR in a few minutes

1 hour later

**Comment by WorldLanguages** [comment]
lol sry, homework happened :P

**Comment by FunctionalMetatable** [comment]
> * Add a "done" button to this modal
> * Close this modal when clicking outside
> * Square pfps and stretch to fit



> * [ ]  Only request followers when modal is active
> * [ ]  Add following (only request when active as well)
> * [ ]  Do infinite scrolling instead of requesting all followers
> * [x]  Alert user if request status is not 200 (but do nothing if we get "this user was already invited" message, which returns 200)
> * [x]  active className

everything is done, hopefully

**Comment by WorldLanguages** [comment]
@mxmou Speak now or forever hold your peace :P 

My plan for today:

- [ ] old-studio-layout should recreate the buttons it moves, and make it so that clicking those buttons opens the bottom bar with the real button on it.
- [ ] Move the "invite followers" button next to the invite by username button (see above for screenshot)
- [ ] Higher z-index for bottom bar in old-studio-layout so that the bottom bar is on top of fixed footer for infinite-scrolling users
- [ ] Fix this addon's manifest

And hopefully merge :P

**Comment by WorldLanguages** [comment]
> Why is the button not next to the invite row?

@mxmou Now that I think about this, that also applies for studio-tools, which adds "promote curators" and "remove curators" below the managers. So for consistency, I'll keep "invite followers" at the top for now.

**Comment by WorldLanguages** [comment]
Merging to get strings into translators - all of my recent changes are reversible

**Comment by mxmou** [comment]
> > Why is the button not next to the invite row?
> 
> @mxmou Now that I think about this, that also applies for studio-tools, which adds "promote curators" and "remove curators" below the managers. So for consistency, I'll keep "invite followers" at the top for now.

It makes sense for "Promote curators" and "Remove curators" to be at the top - promoting curators adds new managers and "Remove curators" can be used to remove both curators and managers.

**Comment by WorldLanguages** [comment]
Description is already localized and says "at the top", so the only option now is to have the button at both places :P

**Pull request merged**